Permission API v1 (#8246)
molgenis/molgenis: docs/developer_documentation/ref-permissions-api.md

# Permission API

This API can be used to perform create, update and delete operations on the MOLGENIS permission system.
The MOLGENIS permission is based on [Spring Security](https://spring.io/projects/spring-security).

## Parameters
These parameters are used by the endpoint in this API:
- 'typeId': This is the type of resource for which the permission is granted. 
This maps to the 'type' in Spring security's [ObjectIdentity](https://docs.spring.io/spring-security/site/docs/4.2.11.RELEASE/apidocs/org/springframework/security/acls/model/ObjectIdentity.html).

> In MOLGENIS #####Examples of these types are 'entityType', 'package', 'plugin', and also row leve secured entities, in which case the type is the entityType identifier perfixed with "-entity" 

Reviewer: There's a header in the middle of the sentence

> In MOLGENIS #####Examples of these types are 'entityType', 'package', 'plugin', and also row leve secured entities, in which case the type is the entityType identifier perfixed with "-entity" 

Reviewer: ```suggestion
In MOLGENIS #####Examples of these types are 'entityType', 'package', 'plugin', and also row level secured entities, in which case the type is the entityType identifier prefixed with "entity-" 
```

- 'identifier': This is the identifier of the actual resource within the resource type for which the permission is granted. 
This maps to the 'identifier' in Spring security's ObjectIdentity.
- 'inheritance': a boolean indicating if inherited permissions should be returned or only the permission that are actually set for the roles and users requested. This parameter is only used in the the GET permission requests. Setting this to true will return a tree with all inherited permissions for the requested users and roles. This field cannot be combined with paging.
- 'page': The pagenumber for the results, should only be provided combined with 'pageSize'. This field cannot be combined with 'inheritance=true'.
- 'pageSize': The number of results per result page, should only be provided combined with 'page'. This field cannot be combined with 'inheritance=true'.
- 'permission': A permission for a resource, the permissions that can be used differ per resource type, but are always a subset of READMETA,COUNT,READ,WRITE,WRITEMETA. 

## Query for user or role
- user

> The user for which to get/create/update the permission. This should be the username as stated in the 'sys_sec_User' table in the MOLGENIS database;

Reviewer: ```suggestion
The user for which to get/create/update the permission. This should be the username as stated in the 'sys_sec_User' table in the MOLGENIS database.
```
Something tells me you're a programmer 😄

- role

> The role for which to get/create/update the permission. This should be the rolename as stated in the 'sys_sec_Role' table in the MOLGENIS database;

Reviewer: ```suggestion
The role for which to get/create/update the permission. This should be the rolename as stated in the 'sys_sec_Role' table in the MOLGENIS database.
```

The user/role query for this API's GET operations should be provided in the [RSQL syntax](developer_documentation/ref-RSQL.md)
Examples:

Query for user 'Cardiologist' or users with role "CARDIOLOGY"
```
q=user==Cardiologist,role==CARDIOLOGY
```
Query for users 'Cardiologist' or 'Neurologist'
```
q=user=in=(Cardiologist,Neurologist)
```
### Permission inheritance
There are two kinds of inheritance in the permission system:
- Users inherit permissions from their roles, and roles from their parentroles.

> - The access control list for a resource can have a parent, in that case permissions on the parent also grant permissions on the child ACL's. In the current MOLGENIS system this is the case for entity types and packages, both inherit permissions from the package in which they reside..

Reviewer: The term ACL hasn't been explained yet.

Warning: this will not create ACL's for already existing resources of this type.
You can use the endpoint in the next section 'Creating a new access control list for a resource' to create ACL's for these resources.

##### Endpoint
```
POST https://molgenis.mydomain.example/api/permissions/v1/types/{typeId}")
```
########## Parameters
URL: 'typeId' as described in the [parameters section](##Parameters)

########## Response
201 CREATED

########## Example
Enable row level security on entity 'hospital_neurology_patients'.

Request:
```https://molgenis.mydomain.example/api/permissions/v1/types/entity-hospital_neurology_patients```

### Creating a new access control list for a resource
##### Endpoint
```
POST https://molgenis.mydomain.example/api/permissions/v1/objects/{typeId}/{objectId}")
```
##### Parameters
URL: 
- 'typeId' as described in the [parameters section](##Parameters)
- 'identifier' as described in the [parameters section](##Parameters)

##### Response
201 CREATED

##### Example
Add an ACL for 'Patient1' in entity 'hospital_neurology_patients'.

Request:
```POST https://molgenis.mydomain.example/api/permissions/v1/types/entity-hospital_neurology_patients/Patient1```

> ### Getting all acls for a class

Reviewer: This is the first time the term `class` is used. Is it used interchangeably with `typeId`?

Reviewer: If it should be called something else than `class`, I think the following error message should also be updated:

```
"message": "Acl class 'entity-it_emx_autoid_testAutoId' already exists.",
"code": "PRM14"
```

Author: I'll take a look, this is an artefact of the renaming. It is caused by the fact that both terms are used in spring security as well...

##### Endpoint
```
GET https://molgenis.mydomain.example/api/permissions/v1/objects/{typeId}
```
##### Parameters
URL:
- 'typeId' as described in the [parameters section](##Parameters)
Query: 
- Optional: 'page' as described in the [parameters section](##Parameters)
- Optional: 'pageSize' as described in the [parameters section](##Parameters)

##### Response
List of ACLs for a class in the system.
##### Example 
Request:
```
GET https://molgenis.mydomain.example/api/permissions/v1/objects/plugin?page=2&pageSize=2
```
Response:
```
{
    "page": {
        "size": 2,
        "totalElements": 36,
        "totalPages": 18,
        "number": 2
    },
    "links": {
        "previous": "/api/permissions/v1/objects/plugin?page=1&pageSize=2",
        "self": "/api/permissions/v1/objects/plugin?page=2&pageSize=2",
        "next": "/api/permissions/v1/objects/plugin?page=3&pageSize=2"
    },
    "data": [
        "background",
        "contact"
    ]
}
   ```

## Retrieving permissions for users and roles

### Getting permissions for one or more users and/or roles for a resource
##### Endpoint
```

> GET https://molgenis.mydomain.example/api/permissions/v1/{typeId}/{objectId}")

Reviewer: ```suggestion
GET https://molgenis.mydomain.example/api/permissions/v1/{typeId}/{objectId}
```

                   }
               ]
           },
           {
               "user": "NeuroNurse",
               "inheritedPermissions": [
                   {
                       "role": "NEUROLOGY",
                       "permission": "READ",
                       "inheritedPermissions": []
                   }
               ]
           },
           {
               "user": "Reception",
               "inheritedPermissions": [
                   {
                       "identifier": "hospital",
                       "label": "hospital",
                       "classLabel": "Package",
                       "classId": "package",
                       "permission": "READ",
                       "inheritedPermissions": []
                   }
               ]
           }
       ]
   }
   ```
The neurologist and the nurse inherit their READ permissions from the their "NEUROLOGY" role, while the reception inherits the READ permission from the READ permission they have on the parent package "hospital" of the "hospital_neurology" package.

### Getting permissions for one or more users and/or roles for a resource type
##### Endpoint
```

> GET https://molgenis.mydomain.example/api/permissions/v1/{typeId}")

Reviewer: ```suggestion
GET https://molgenis.mydomain.example/api/permissions/v1/{typeId}
```

                            ]
                        }
                    ]
                }
            ]
        },
        {
            "classId": "plugin",
            "label": "Plugin",
            "rowPermissions": [
                {
                    "identifier": "dataexplorer",
                    "label": "dataexplorer",
                    "permissions": [
                        {
                            "role": "CARDIOLOGY",
                            "permission": "READ",
                            "inheritedPermissions": []
                        }
                    ]
                },
                {
                    "identifier": "navigator",
                    "label": "navigator",
                    "permissions": [
                        {
                            "role": "CARDIOLOGY",
                            "permission": "READ",
                            "inheritedPermissions": []
                        }
                    ]
                }
            ]
        }
    ]
}
```

## Granting permissions to users and roles

### Creating permissions for one or more users and/or roles for a resource
##### Endpoint
```

> POST https://molgenis.mydomain.example/api/permissions/v1/{typeId}/{objectId}")

Reviewer: ```suggestion
POST https://molgenis.mydomain.example/api/permissions/v1/{typeId}/{objectId}
```

201 CREATED

##### Example request
URL:
https://molgenis.mydomain.example/api/permissions/v1/entityType/hospital_cardiology_patients
Body:
```{
   	permissions:[{
   		permission:READ,
   		role:CARDIOLOGY
   	},{
   		permission:WRITE,
   		user:Cardiologist
   	}]
   }
   ```

### Create permissions for one or more users and/or roles for resources of a certain type
```
POST https://molgenis.mydomain.example/api/permissions/v1/{typeId}")
``` 

##### Request
The endpoint expects a list of resources, each of which should contrain the identifier for the resource and a list of permissions, each of these permission should contain a 'permission' and a 'user' or a 'role'.

##### Response 
201 CREATED

##### Example
URL: https://molgenis.mydomain.example/api/permissions/v1/entity-hospital_neurology_patients
Body:
```
{
	rows:[{
			identifier:Patient1,
			permissions:[
			{
				role:CARDIOLOGY,
				permission:WRITE
			}
		]
	},{
			identifier:Patient2,
			permissions:[
			{
				user:Cardiologist,
				permission:WRITE
			},
			{
				user:CardioNurse,
				permission:READ
			}
		]}
	]
}
```

### Update permissions for one or more users and/or roles for a resource type
##### Endpoint
```

> PATCH https://molgenis.mydomain.example/api/permissions/v1/{typeId}/{objectId}")

Reviewer: ```suggestion
PATCH https://molgenis.mydomain.example/api/permissions/v1/{typeId}/{objectId}
```

```
Request: 
The endpoint expects a list of permissions, each permission should contain a 'permission' and a 'user' or a 'role'.

##### Response
200 OK

##### Example 
Request:
URL: https://molgenis.mydomain.example/api/permissions/v1/entityType/hospital_cardiology_patients
Body:
```{
   	permissions:[{
   		permission:WRITE,
   		role:CARDIOLOGY
   	},{
   		permission:WRITEMETA,
   		user:Cardiologist
   	}]
   }
   ```

### Update permissions for one or more users and/or roles for resources of a certain type
```

> PATCH https://molgenis.mydomain.example/api/permissions/v1/{typeId}")

Reviewer: ```suggestion
PATCH https://molgenis.mydomain.example/api/permissions/v1/{typeId}
```

``` 

##### Request
The endpoint expects a list of resources, each of which should contrain the identifier for the resource and a list of permissions, each of these permission should contain a 'permission' and a 'user' or a 'role'.

##### Response 
200 OK

##### Example
URL: https://molgenis.mydomain.example/api/permissions/v1/entity-hospital_neurology_patients
Body:
```
{
	rows:[{
			identifier:Patient1,
			permissions:[
			{
				role:CARDIOLOGY,
				permission:WRITE
			}
		]
	},{
			identifier:Patient2,
			permissions:[
			{
				user:Cardiologist,
				permission:WRITE
			},
			{
				user:CardioNurse,
				permission:READ
			}
		]}
	]
}
```

### Removing permissions for one or more users and/or roles for a resource

##### Endpoint
```

> DELETE https://molgenis.mydomain.example/api/permissions/v1/{typeId}/{objectId}")

Reviewer: ```suggestion
DELETE https://molgenis.mydomain.example/api/permissions/v1/{typeId}/{objectId}
```

        "label": "Package"
    },
    {
        "id": "entity-sys_FileMeta",
        "entityType": "sys_FileMeta",
        "label": "File metadata"
    },
    {
        "id": "plugin",
        "entityType": "sys_Plugin",
        "label": "Plugin"
    },
    {
        "id": "entityType",
        "entityType": "sys_md_EntityType",
        "label": "Entity type"
    }
]
   ```

### Getting all suitable permissions for a resource type
##### Endpoint
```
GET https://molgenis.mydomain.example/api/permissions/types/permissions/{typeId}
```
##### Parameters
URL: TypeId as described in the [parameters section](##Parameters)

##### Response
| Status code         | Description                                                                       |
|---------------------|-----------------------------------------------------------------------------------|
|200 Ok | Success |
|403 Forbidden | the user has no permissions to execute this action |
|404 Not found | the type does not exist |

##### Response body
A list of permissions that can be used for this resource type.

##### Example 
Request:
```https://molgenis.mydomain.example/api/permissions/types/permissions/entityType```
Response:

> ```[

Reviewer 2: See previous comment regarding https://opensource.zalando.com/restful-api-guidelines/#110.

       "READMETA",
       "READ",
       "COUNT",
       "WRITEMETA",
       "WRITE"
   ]
   ```

> ### Creating a new resource type in the system (Row level securing an entity)

Reviewer 2: entity -> entity type

> ### Creating a new resource type in the system (Row level securing an entity)

Reviewer 2: add: feature enabled only for superusers

Reviewer 2: Disabling: ```This will delete a type, the access control lists for all rows will remain in the system.```. Could it be that enabling uses this information and only sets the current user as owner for rows that were added after disabling?

This will create a new type in the system and add access control lists for all existing rows.
This enables row level security.
Please keep in mind that all existing rows will get the current user as owner.

##### Endpoint
```
POST https://molgenis.mydomain.example/api/permissions/types/{typeId}
```
##### Parameters
URL: 'typeId' as described in the [parameters section](##Parameters)

##### Response
| Status code         | Description                                                                       |
|---------------------|-----------------------------------------------------------------------------------|
|201 Created | Type created |
|403 Forbidden  | the user has no permissions to execute this action |
|409 Conflict | the type already exists |

##### Example

> Enable row level security on entity 'hospital_neurology_patients'.

Reviewer 2: entity -> entity type

Request:
```https://molgenis.mydomain.example/api/permissions/types/entity-hospital_neurology_patients```

> ### Deleting a resource type from the system (Removing row level security from an entity)

Reviewer 2: entity -> entity type

This will delete a type, the access control lists for all rows will remain in the system.
This disables row level security.

##### Endpoint
```
POST https://molgenis.mydomain.example/api/permissions/types/{typeId}
```
##### Parameters
URL: 'typeId' as described in the [parameters section](##Parameters)

##### Response
| Status code         | Description                                                                       |
|---------------------|-----------------------------------------------------------------------------------|
|204 No content| Type removed |
|403 Forbidden  | the user has no permissions to execute this action |
|404 Not found | the type does not exist |

##### Example

> Enable row level security on entity 'hospital_neurology_patients'.

Reviewer 2: entity -> entity type

Request:
```https://molgenis.mydomain.example/api/permissions/types/entity-hospital_neurology_patients```

### Creating a new access control list for a resource
##### Endpoint
```
POST https://molgenis.mydomain.example/api/permissions/objects/{typeId}/{objectId}
```
##### Parameters
URL: 
- 'typeId' as described in the [parameters section](##Parameters)
- 'identifier' as described in the [parameters section](##Parameters)

##### Response
| Status code         | Description                                                                       |
|---------------------|-----------------------------------------------------------------------------------|
|201 Created | access control list created |
|400 Bad request| the permission request content is not valid |
|403 Forbidden  | the user has no permissions to execute this action |
|404 Not found | the type, object, user, role does not exist |
|409 Conflict | the a access control list already exists for this object |

##### Example

> Add an ACL for 'Patient1' in entity 'hospital_neurology_patients'.

Reviewer 2: entity -> entity type
